docs/source/daml/reference/contract-keys.rst

```diff
@@ -0,0 +1,95 @@
+
+Checking of the keys is done automatically at execution time, by the DAML exeuction engine: if someone tries to create a new contract that duplicates an existing contract key, the execution engine will cause that creation to fail. 
+
+Contract keys functions
+***********************
+
+Contract keys introduce several new functions.
+
+``fetchByKey``
+==============
+
+``fetchedContract <- fetchByKey @ContractType contractKey``
+
+Use ``fetchByKey`` to fetch the ID and data of the contract with the specified key. It is an alternative to the currently-used ``fetch``.
+
+It returns a tuple of the ID and the contract object (containing all its data). 
+
+You need authorization from **at least one** of the maintainers to run ``fetchByKey``. A maintainer can authorize by being a signatory, or by submitting the command/being a controller for the choice.
+
+``fetchByKey`` fails and aborts the transaction if:
+
+- you don't have sufficient authorization
+- you're not a stakeholder of the contract you're trying to fetch
+
+This means that if it fails, it doesn't guarantee that a contract with that key doesn't exist, just that you can't see one.
+
+Because the type is ambiguous, when calling you need to specify what you expect with ``@`` and the type, for example ``@MyTemplateType``.
+
+``lookupByKey``
+===============
+
+``contractId <- lookupByKey @ContractType contractKey``
+
+Use ``lookupByKey`` to check whether a contract with the specified key exists. If it does exist, ``lookupByKey`` returns the ``ContractId`` of the contract; otherwise, it returns ``None``.
+
+You need authorization from **all** of the maintainers to run ``lookupByKey``, and it can only be submitted by one of the maintainers.
+
+If the lookup fails (ie returns ``None``), this guarantees that no contract has this key.
+
+Unlike ``fetchByKey``, the transaction **does not fail** if a contract with the key doesn't exist: instead, ``lookupByKey`` just returns ``None``.
+
+To get the data from the contract once you've confirmed it exists, you'll still need to use ``fetch``.
+
+Because the type is ambiguous, when calling you need to specify what you expect with ``@`` and the type, for example ``@MyTemplateType``.
+
+``exerciseByKey``
+=================
+
+``exerciseByKey @ContractType contractKey``
+
+Use ``exerciseByKey`` to exercise a choice on a contract identified by its ``key`` (compared to ``exercise``, which lets you exercise a contract identified by its ``ContractId``).
+
+Because the type is ambiguous, when calling you need to specify what you expect with ``@`` and the type, for example ``@MyTemplateType``.
+
+Error messages
+**************
+
+If you don't include the ``maintainer`` in your ``key``, you'll see the following error::
+
+   Failure to process DAML program, this feature is not currently supported.
+   Unbound reference to this in maintainer with evar.
```

Comment:
```suggestion
   Unbound reference to this in maintainer with evar.
   
We plan to improve the error messages in the near future.
```

Reply:
Nope - I really don't like committing to future improvements in the docs. Doesn't help the users at the point they read it, just makes us feel better.

Reply:
Oops, the "We plan to improve the error messages in the near future." was not meant to be part of the suggestion.